storage/cache/map_subtree_cache.go

```diff
@@ -39,6 +40,7 @@ func NewMapSubtreeCache(mapStrata []int, treeID int64, hasher hashers.MapHasher)
 func populateMapSubtreeNodes(treeID int64, hasher hashers.MapHasher) storage.PopulateSubtreeFunc {
 	return func(st *storagepb.SubtreeProto) error {
 		st.InternalNodes = make(map[string][]byte)
+		var internalNodesMu sync.Mutex
```

Comment:
I _suspect_ there's probably not that much benefit in making the population of internal subtree nodes concurrent (given there'll be plenty more "real" concurrency at the next level up where disk/DB accesses are happening).

I wonder if it's worth making `NewHStar2` take some sort of concurrency parameter?

Reply:
That may well be true, but HStar2 isn't aware of what nodes are in the cache vs. going to result in a db access, and I'm not sure it would be a good idea to give it that education for layering reasons. 

I did think about doing something like worker pools for HStar2 to handle the massive go routine fanout, but worker pools can easily get into a deadlock situation.  In this configuration where threads will block on other threads, we need the more intelligent block/unblocking logic that the native goroutines provide.

Reply:
> HStar2 isn't aware of what nodes are in the cache vs. going to result in a db access

Exactly - but the caller may have more context - like this site where `hstar2` is being used only to populate the subtree cache structure...

Reply:
Yes, I think there's hardly a point in populating the cache on **writes** concurrently. But there are also **reads** which need to be sped up: either by batching (#1405 basically achieves the same effect one layer above because it's aware of the cache), or parallelization (this PR). I was starting a PR refactoring the `HStar2` itself, so that it would return a list of nodes to fetch, without relying on the cache.

Reply:
That refactor is possibly pretty worthwhile - the way the hstar2, cache, and storage interconnect is a little complex as it - pulling it apart to: 
   1. figure out pages we need
   2. load the pages
   3. run hstar2 with *only* access to the loaded pages
   4. flush writes

would be much easier to grok.

There might be some performance improvement from parallelisation, although in the current state there are already 256 goroutines servicing the shards, but with Circus Tent it might be helpful to be able to run sub-shards within each shard server concurrently too.

Reply:
I agree. The step 1 will probably decompose into two (processed at different levels):
1.1. Figure out **nodes** that HStar2 needs
1.2. Transform it to the list of **pages** to load